### ISSUE TITLE
OpenEntityManagerInViewInterceptor never closes connections when `spring.mvc.async.request-timeout` is set to `-1`

### ISSUE THREAD

**Issue by rburgst** [spring-projects/spring-boot#12110]
Spring Boot: ´1.5.9.RELEASE´

I have an SSE application that uses hibernate and spring-data. Therefore `OpenEntityManagerInViewInterceptor` is automatically set up for me by spring boot autoconfig.

In order to reduce useless error logs I found https://github.com/codecentric/spring-boot-admin/issues/317 and configured `spring.mvc.async.request-timeout: -1`.

This configuration (however) leads to the fact that the connections are never closed because (most likely) in `org.springframework.web.servlet.DispatcherServlet#processDispatchResult` 

```

		if (WebAsyncUtils.getAsyncManager(request).isConcurrentHandlingStarted()) {
			// Concurrent handling started during a forward
			return;
		}

		if (mappedHandler != null) {
			mappedHandler.triggerAfterCompletion(request, response, null);
		}
```

we use the `isConcurrentHandlingStarted()` branch and therefore the `triggerAfterCompletion` method is not being called.

Making matters (even) worse is the default configuration for the tomcat connection pool where by default `removeAbandoned=false`. 
Basically this leads to the connection pool slowly draining all its connections until no connections are available anymore.

**Comment by wilkinsona** [comment]
Thanks for the report. From what you've said, it sounds like this is working as I'd expect. If you start an async request, never end it, and disable timeout, then it'll last indefinitely. Perhaps you were expecting after completion to be triggered prior to the async dispatch? If so, that would require a change to Spring Framework (and I'm not sure it makes sense for the general interceptor case). So that we can triage this effectively, can you please provide a [complete, minimal example](https://stackoverflow.com/mcve) that illustrates the problem?

**Comment by rburgst** [comment]
Sure thing, I will wip something together in the next hour or so.

**Comment by rburgst** [comment]
here is a sample repo
https://github.com/rburgst/sse-jpa-demo
the instructions are shown in the readme. 

The twist is that connections only get closed when there is a problem on the SSE sending, if I dont send anything for a long time, then the connections will leak.

Maybe I am holding it wrong, but while the code is a bit naiive, noone really would think about who will cleanup the DB connections in such a case. I guess this is the curse of spring handling everything for you. 
So lets just assume for now that it works as intended. How would one properly manage the Hibernate session in such a case (e.g. in the `register` method we need to load the current status from the DB and all other changes are done incrementally without needing the DB).

**Comment by wilkinsona** [comment]
Thanks for the excellent sample. It's confirmed my suspicion that is is working is designed. In short, I don't think that using OEMIV interceptor with async request handling and no timeout is a good idea. As you have observed, as the long-running async requests mount up, so do the open entity managers.

Your sample doesn't make use of the entity manager during view rendering, so you can avoid the problem by disabling the interceptor:

```
spring.jpa.open-in-view=false
```

If your actual application is more complex and was relying on the interceptor, then you'll have to update it to avoid any `LazyInitializationException`s that may now be thrown.

/cc @rstoyanchev in case I've got anything wrong about OEMIVI.

**Comment by rburgst** [comment]
Just out of curiosity, how would I manually manage the `EntityManager` in case I needed the OEMIV interceptor in a way that wouldn't conflict with the interceptor?

**Comment by wilkinsona** [comment]
It's very hard to say without knowing more about your application, and a GitHub issue isn't really the right place for that sort of discussion. One possibility may be to create and register a custom OpenEntityManagerInViewInterceptor subclass that ignores certain requests. If you need some further guidance please follow up on Stack Overflow or Gitter.

**Comment by rstoyanchev** [comment]
In async request scenario, `afterCompletion` is not invoked until the entire async request completes, see the Javadoc of `AsyncHandlerInterceptor` sub-interface.

> Just out of curiosity, how would I manually manage the EntityManager in case I needed the OEMIV interceptor in a way that wouldn't conflict with the interceptor?

When the async request completes -- via timeout, explicit completion, or an error, a final dispatch will be made to complete async processing, at which point `afterCompletion` will be called. In other words you shouldn't have to manually manage the `EntityManager`. Just make sure that the async request completes one way or another.

`EntityManager` aside, you have to either set a timeout or emit periodically, or otherwise you wouldn't detect disconnected clients (the Servlet API does not provide a notification if a client goes away) which would drain your Servlet container connection pool too.
